sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala

```
@@ -61,25 +65,50 @@ import org.apache.spark.sql.catalyst.plans.logical.LogicalKeyedState
+ *    - Timeout will never be called before real clock time has advanced by D ms
+ *    - Timeout will be called eventually when there is a trigger in the query
+ *      (i.e. after D ms). So there is a no strict upper bound on when the timeout would occur.
+ *      For example, the trigger interval of the query will affect when the timeout is actually hit.
+ *      If there is no data in the stream (for any key) for a while, then their will not be
+ *      any trigger and timeout will not be hit until there is data.
+ *
  * Scala example of using KeyedState in `mapGroupsWithState`:
  * {{{
  * // A mapping function that maintains an integer state for string keys and returns a string.
  * def mappingFunction(key: String, value: Iterator[Int], state: KeyedState[Int]): String = {
- *   // Check if state exists
- *   if (state.exists) {
- *     val existingState = state.get  // Get the existing state
- *     val shouldRemove = ...         // Decide whether to remove the state
+ *
+ *   if (state.isTimingOut) {                // If called when timing out, remove the state
+ *     state.remove()
+ *
+ *   } else if (state.exists) {              // If state exists, use it for processing
+ *     val existingState = state.get         // Get the existing state
+ *     val shouldRemove = ...                // Decide whether to remove the state
  *     if (shouldRemove) {
- *       state.remove()     // Remove the state
+ *       state.remove()                      // Remove the state
+ *
  *     } else {
  *       val newState = ...
- *       state.update(newState)    // Set the new state
+ *       state.update(newState)              // Set the new state
  *     }
+ *
  *   } else {
  *     val initialState = ...
- *     state.update(initialState)  // Set the initial state
+ *     state.update(initialState)            // Set the initial state
  *   }
- *   ... // return something
+ *   state.setTimeoutDuration("1 hour")      // Set the timeout
```

Hmm, I wonder if we should throw an exception here.  I was pretty confused what this function would do just looking at it.